miner/agent.go

```diff
@@ -58,8 +61,12 @@ func (self *CpuAgent) Stop() {
 }
 
 func (self *CpuAgent) Start() {
-	self.mu.Lock()
 	defer self.mu.Unlock()
+	self.mu.Lock()
+
+	if atomic.LoadInt32(&self.isMining) == 1 {
+		return // agent already started
+	}
```

Comment:
I'm aware of the CAS paradigm. In this case it makes the code much harder to read with only a very small performance gain. I don't think its worth it.